### PR TITLE
[serve] gRPC proxy to handle internal errors

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Found this unhandled case when writing the doc. The current behavior did not send gRPC "INTERNAL" status back to the client. For unary requests, it will just fail silently without the client getting the error. For streaming requests, although the errors are returned, the status code are not properly propagated to the client. This PR ensure the RPC calls set the expected status code and error messages when there are internal error raised. Also added a test case for it to ensure it's caught moving forward. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/http_proxy.py [hunk 1]
```diff
@@ -703,6 +703,14 @@ async def health_response(self, proxy_request: ProxyRequest) -> ProxyResponse:
             response=response_proto.SerializeToString(),
         )
 
+    async def internal_error_response(
+        self, proxy_request: ProxyRequest, error: Exception
+    ) -> ProxyResponse:
+        status_code = grpc.StatusCode.INTERNAL
+        proxy_request.send_status_code(status_code=status_code)
+        proxy_request.send_details(message=str(error))
+        return ProxyResponse(status_code=str(status_code))
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:712]: why's this async def if it doesn't call any coroutines?

also does this mean that `send_status_code` and `send_details` can block the event loop?

Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:712]: Thanks for discussion offline! Will drop async def as it doesn't do anything in this case. 

Also checked [doc](https://grpc.github.io/grpc/python/grpc.html#grpc.ServicerContext.set_code) and [implementation](https://github.com/grpc/grpc/blob/f5e02f6c620f0a9802cc72ae4ce888e184b06c91/src/python/grpcio/grpc/_server.py#L410-L419) for `set_code` and `set_details`. They just setting an attribute on the request context, doesn't involve network or heavy compute. Those are non-blocking calls 😄

File: python/ray/serve/_private/http_proxy.py [hunk 2]
```diff
@@ -821,6 +829,9 @@ async def _streaming_generator_helper(
 
             except StopAsyncIteration:
                 break
+            except Exception as e:
+                self.internal_error_response(proxy_request, e)
```
Comment (sihanwang41) [inline python/ray/serve/_private/http_proxy.py:833]: nit: set_internal_error_response()

File: python/ray/serve/_private/http_proxy.py [hunk 3]
```diff
@@ -703,6 +703,14 @@ async def health_response(self, proxy_request: ProxyRequest) -> ProxyResponse:
             response=response_proto.SerializeToString(),
         )
 
+    def internal_error_response(
```
Comment (sihanwang41) [inline python/ray/serve/_private/http_proxy.py:706]: should this function being a util func?  look like not related to the instance of `gRPCProxy`.

Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:706]: I think this should stay within `gRPCProxy` it can only be used when responding gRPC request as well. But will also give it a `_` :)

**Comment by GeneDer** [comment]
@edoakes @sihanwang41 Just FYI I addressed the comments. PTAL when you have a sec, would be nice if we can get this cherry-pick today🙏

**Review by edoakes — APPROVED** [review]
